### PR TITLE
Cleanup device plugin - first round

### PR DESCRIPTION
**Pull request by tengqm**
**What this PR does / why we need it**:
Clean up device plugin implementation.
- remove locks which is obviously unecessary
- add/fix function comments
- move reconcile function out
- add TODOs
 
**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/kubelet/cm/deviceplugin/endpoint.go [hunk 1]
```diff
@@ -36,7 +37,7 @@ type endpoint interface {
 	run()
 	stop()
 	allocate(devs []string) (*pluginapi.AllocateResponse, error)
-	getDevices() []pluginapi.Device
+	getDevices() map[string]pluginapi.Device
```
Comment (RenaudWasTaken) [inline pkg/kubelet/cm/deviceplugin/endpoint.go]: Any reason for making this a map ?

File: pkg/kubelet/cm/deviceplugin/endpoint.go [hunk 2]
```diff
@@ -77,24 +78,22 @@ func (e *endpointImpl) callback(resourceName string, added, updated, deleted []p
 	e.cb(resourceName, added, updated, deleted)
 }
 
-func (e *endpointImpl) getDevices() []pluginapi.Device {
+// getDevices makes a deep copy of the ID->Device mapping
+func (e *endpointImpl) getDevices() map[string]pluginapi.Device {
```
Comment (RenaudWasTaken) [inline pkg/kubelet/cm/deviceplugin/endpoint.go]: I've already addressed cleanup of the Device interface in PR #52616

**Review by RenaudWasTaken — CHANGES_REQUESTED** [review]
Do you mind removing the endpoint management part?
Would like to have comments on PR #52616 for changes related to endpoints.

Everything else lgtm

File: pkg/kubelet/cm/deviceplugin/endpoint.go (hunk 1 above)
Comment (tengqm) [reply pkg/kubelet/cm/deviceplugin/endpoint.go]: It took me quite some time to figure out that this needs a change.

In the previous implementation, getDevices() return a list/slice. If you search the code for invocations of this method, you can see that only the Manager interface wants a list. In all other cases, a map is needed.
There was even a case where we duplicate this logic with another mutex lock:

```
  func (e) run() {
       devices := make(map ...)
       e.mutex.lock()
       for _, d := range.devices {
          devices[d.id] = d
       }
      e.mutex.unlock()
```

After making this change, the only tax added is the Manager.Devices() call, which has to translate the map to a slice. All other cases when we need the device map, we just call getDevices().

So ... it saves the unnecessary back-forth conversion, it saves unnecessary locks and simpifies the code.

**Comment by lichuqiang** [comment]
I have several ideas for refactoring too, but I think we'd better wait for #52616 in, and then do some refactoring/cleanup basing on a relatively stable architecture

File: pkg/kubelet/cm/deviceplugin/endpoint.go (hunk 2 above)
Comment (tengqm) [reply pkg/kubelet/cm/deviceplugin/endpoint.go]: Right, I should have spent some time on that PR. I went sleepless last night thinking of ways to get the concurrency management do right. The data structure has to be correct or else all the efforts on getting mutex/rwmutex done right will go in vain.

**Comment by tengqm** [comment]
@lichuqiang Great. Let's work together to make it a beta feature in v1.10. :)

**Comment by RenaudWasTaken** [comment]
While this PR, I've noticed that the `TestCheckpoint` doesn't cleanup the checkpoint file.
Do you mind fixing it here ?

File: pkg/kubelet/cm/deviceplugin/endpoint.go (hunk 1 above)
Comment (RenaudWasTaken) [reply pkg/kubelet/cm/deviceplugin/endpoint.go]: It makes sense, I've actually unconsciously done the same thing in #52616 but masked it behind an interface.

File: pkg/kubelet/cm/deviceplugin/endpoint.go (hunk 2 above)
Comment (RenaudWasTaken) [reply pkg/kubelet/cm/deviceplugin/endpoint.go]: Agreed current concurrency management is a bit nuts...

**Comment by RenaudWasTaken** [comment]
Also while we are at it probably remove [this nil check on a list](https://github.com/kubernetes/kubernetes/blob/45db5e7260d47f01106c3f0401f9c779f9b386c0/pkg/kubelet/kubelet_node_status.go#L605) before iterating over it

**Comment by RenaudWasTaken** [comment]
[Use `nil` rather than `[]pluginapi.Device{}`](https://github.com/kubernetes/kubernetes/blob/45db5e7260d47f01106c3f0401f9c779f9b386c0/pkg/kubelet/cm/deviceplugin/manager_test.go#L173). But this sounds more like personal coding style

**Comment by RenaudWasTaken** [comment]
[Use `assert.Len` rathen than `assert.Equal`](https://github.com/kubernetes/kubernetes/blob/45db5e7260d47f01106c3f0401f9c779f9b386c0/pkg/kubelet/cm/deviceplugin/manager_test.go#L177) could also be a good thing.

**Comment by tengqm** [comment]
@RenaudWasTaken I haven't touched the unit tests yet. I'm currently looking into the manager layer since the endpoint/device layer concurrency is being fixed in #52616 . My intention is to combine allDevices, allocatedDevices and poddevices, or at least reduce the copying. I'm not sure if it is possible to split the locks. Please let me know if there are already efforts on this.

**Comment by RenaudWasTaken** [comment]
@tengqm I'm not aware of any other efforts to do this (other than yours).

**Comment by tengqm** [comment]
@RenaudWasTaken endpoint test case fixed. e2e gpu test failure is irrelevant.

**Comment by tengqm** [comment]
Closing this in favor of #56398.

**Pull request closed without merging**